### PR TITLE
Allow for users to declare explicit build timestamps

### PR DESCRIPTION
**Pull request by dekkagaijin**
Fixes: #155 

Signed-off-by: Jake Sanders <jsand@google.com>

### CODE REVIEW

**Comment by dekkagaijin** [comment]
Can do the `stamp=True` implementation, as well, but didn't know if there was one preference over the other.

File: container/create_image_config.py [hunk 1]
```diff
@@ -121,10 +125,30 @@ def Stamp(inp):
...
   output = v2_2_metadata.Override(data, v2_2_metadata.Overrides(
-      author='Bazel', created_by='bazel build ...',
-      layers=layers, entrypoint=list(map(Stamp, fix_dashdash(args.entrypoint))),
-      cmd=list(map(Stamp, fix_dashdash(args.command))), user=Stamp(args.user),
+      author='Bazel',
+      created_by='bazel build ...',
+      layers=layers,
+      entrypoint=list(map(Stamp, fix_dashdash(args.entrypoint))),
+      cmd=list(map(Stamp, fix_dashdash(args.command))),
+      creation_time=creation_time,
```
Comment (mattmoor) [inline container/create_image_config.py]: This doesn't fix the issue unless users can use: `{SOURCE_DATE_EPOCH}` (or w/e the magic stamp variable @ixdy wants is) for this value.

File: testdata/BUILD [hunk 2]
```diff
@@ -244,6 +244,18 @@ container_image(
...
+container_image(
+    name = "with_unix_epoch_creation_time",
+    base = ":base_with_volume",
+    creation_time = "610203480",
+)
+
+container_image(
+    name = "with_rfc_3339_creation_time",
+    base = ":base_with_volume",
+    creation_time = "1989-05-03T12:58:00Z",
```
Comment (mattmoor) [inline testdata/BUILD]: This kwarg doesn't feel particularly valuable unless its value can be `{SOURCE_DATE_EPOCH}`.  In fact, I wasn't even imagining exposing this, just making `{SOURCE_DATE_EPOCH}` the default when stamping is enabled.

**Comment by mattmoor** [comment]
Thanks for picking this up.

File: testdata/BUILD (hunk 2 above)
Comment (dekkagaijin) [reply testdata/BUILD]: My understanding was that {SOURCE_DATE_EPOCH} would be set as an environment variable, which could be substituted in the build attributes, but I'd believe that I misunderstood the conversation in #155 .

I'll do the `stamp`-based solution.

Comment (dekkagaijin) [reply testdata/BUILD]: I was imagining that people could define their own values, or use `--define` to inject their own values here, without going through `stamp = True` and `SOURCE_DATE_EPOCH`.

File: container/create_image_config.py (hunk 1 above)
Comment (dekkagaijin) [reply container/create_image_config.py]: They can, kind of, though `$(SOURCE_DATE_EPOCH)`  is seconds since the unix epoch and BUILD_TIMESTAMP is in milliseconds.

**Comment by dekkagaijin** [comment]
De nada...

Anyway, done. FYI `bazel test` and `bazel build` have started failing locally within the last week.

File: container/create_image_config.py (hunk 1 above)
Comment (ixdy) [reply container/create_image_config.py]: right, as I understand it, `BUILD_TIMESTAMP` is normally set to the current time, but if `SOURCE_DATE_EPOCH` is set, Bazel sets `BUILD_TIMESTAMP` to that value instead.

tl;dr I think this is doing what we want

File: testdata/BUILD (hunk 2 above)
Comment (ixdy) [reply testdata/BUILD]: maybe also update the `README.md` describing the `creation_time` arg?

File: container/create_image_config.py (hunk 1 above)
Comment (ixdy) [reply container/create_image_config.py]: (it's confusing because they're both yelling case, but `SOURCE_DATE_EPOCH` is an environment variable Bazel recognizes, but it isn't set in the workspace status; only `BUILD_TIMESTAMP` is (unless you set it yourself))

File: testdata/BUILD [hunk 3]
```diff
@@ -244,6 +244,25 @@ container_image(
...
+)
+
+container_image(
+    name = "with_stamped_creation_time",
+    base = ":base_with_volume",
+    creation_time = "{BUILD_TIMESTAMP}",
```
Comment (ixdy) [inline testdata/BUILD]: I sort of wonder whether we want to allow the flexibility of choosing a particular workspace status variable, or if we should just always use `BUILD_TIMESTAMP` when stamping is enabled.

Comment (mattmoor) [reply testdata/BUILD]: That's kind of what I had in mind as well.

Comment (ixdy) [reply testdata/BUILD]: `SOURCE_DATE_EPOCH` is the standard proposed by Reproducible Builds: https://reproducible-builds.org/specs/source-date-epoch/

It makes sense to me to standardize on `BUILD_TIMESTAMP` since that's likely what other rules might use as well.

Comment (ixdy) [reply testdata/BUILD]: I feel that giving the option to specify a different workspace status variable is unnecessary.

I don't have a strong opinion about allowing hard-coded values here, though at least for Kubernetes we plan to just set `SOURCE_DATE_EPOCH` for reproducibility (and cacheability).

Comment (mattmoor) [reply testdata/BUILD]: I'm fine with allowing users to set this, but let's default to `{BUILD_TIMESTAMP}` when `stamp=True` so that they don't have to?

Comment (dekkagaijin) [reply testdata/BUILD]: Works for me. Done.

File: testdata/BUILD (hunk 2 above)
Comment (dekkagaijin) [reply testdata/BUILD]: @ixdy Done.

**Comment by dekkagaijin** [comment]
@mattmoor ping <3

File: container/create_image_config.py [hunk 4]
```diff
@@ -121,10 +126,34 @@ def Stamp(inp):
     elif '{' in value:
       labels[label] = Stamp(value)
 
+  creation_time = None
+  if args.creation_time:
+    creation_time = Stamp(args.creation_time)
+    try:
+      # If creation_time is parsable as an int type, assume unix epoch timestamp.
```
Comment (ixdy) [inline container/create_image_config.py]: oh good, you handled the breaking behavior soon to arrive in bazel (https://bazel-review.googlesource.com/c/bazel/+/48211).

maybe add a TODO to fix this before 2038. :)

Comment (dekkagaijin) [reply container/create_image_config.py]: T_T
I'll wait until the change lands

Comment (dekkagaijin) [reply container/create_image_config.py]: Should I make this change require bazel 0.12? Or should I just create a reasonable test to support both, e.g. parseable as float -> unix timestamp, >1e10 -> milliseconds, <= 1e10 -> seconds

Comment (ixdy) [reply container/create_image_config.py]: I'd vote for "gracefully" handling both forms of timestamps, rather than requiring bazel 0.12+.

Comment (dekkagaijin) [reply container/create_image_config.py]: Done.

**Comment by dekkagaijin** [comment]
@mattmoor ping now that you're back from vacation
